### PR TITLE
Split into separate files and add unit tests

### PR DESCRIPTION
**Pull request by jcansdale**
A while back I was doing some work on VsixUtil, but finding it tricky to change stuff without worrying about breaking it. I ended up splitting the source into separate files and adding a bunch of unit tests. I realize this is a big change, so feel free to tell me where to stick my PR. ;-)

If you were to accept it, it would certainly make it easier to submit PRs in future.

Do you still use this utility yourself at all?

Cheers,
Jamie.

### CODE REVIEW

File: Src/Program.cs [hunk 1]
```diff
@@ -1,490 +1,20 @@
...
             if (args.Length == 0)
             {
                 return CommandLine.Help;
             }
 
             var toolAction = ToolAction.Help;
-            var version = "";
-            var sku = "Community;Professional;Enterprise";
+            VsVersion? version = null;
+            string product = null;
```
Comment (jcansdale) [inline Src/Program.cs:17]: It will now install on all SKUs by default.

File: Src/VsVersion.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace VsixUtil
+{
+    public enum VsVersion
```
Comment (grokys) [inline Src/VsVersion.cs]: Hey Jamie, apologies for the noise. Testing PR comments.

File: Src/Program.cs [hunk 3]
```diff
@@ -497,7 +27,7 @@ private static CommandLine ParseCommandLine(string[] args)
                     case "/install":
                         if (index + 1 >= args.Length)
                         {
-                            Console.Write("/install requires an argument");
+                            consoleContext.Write("/install requires an argument");
```
Comment (jcansdale) [inline Src/Program.cs:30]: This makes it easy to test console output.

File: Src/Program.cs [hunk 4]
```diff
@@ -566,84 +96,69 @@ private static CommandLine ParseCommandLine(string[] args)
...
+        }
+
+        internal static void PrintHelp(IConsoleContext consoleContext)
+        {
+            consoleContext.WriteLine("vsixutil [/install] extensionPath [/rootSuffix name]");
+            consoleContext.WriteLine("vsixutil /uninstall identifier [/rootSuffix name]");
+            consoleContext.WriteLine("vsixutil /list [filter]");
+            consoleContext.WriteLine("");
+            consoleContext.WriteLine("The following filters can be used with all commands:");
+            consoleContext.WriteLine("   /version 15 | 2017");
```
Comment (jcansdale) [inline Src/Program.cs:160]: Version can be specified release number or a year.

File: Src/Program.cs [hunk 5]
```diff
@@ -566,84 +96,69 @@ private static CommandLine ParseCommandLine(string[] args)
...
+            consoleContext.WriteLine("   /product com | Community");
```
Comment (jcansdale) [inline Src/Program.cs:161]: Whole or partial product SKUs can be specified (e.g. com, ent, pro or team).

**Comment by jcansdale** [comment]
Not sure why the build is broken. Does AppVeyor not automatically restore NuGet packages?

File: Test/VsixUtil.Tests/packages.config [hunk 6]
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Microsoft.VisualStudio.Setup.Configuration.Interop" version="1.8.24" targetFramework="net452" developmentDependency="true" />
```
Comment (jcansdale) [inline Test/VsixUtil.Tests/packages.config:3]: I'm not sure why this isn't being restored in CI?

**Comment by jaredpar** [comment]
@jcansdale 

> Not sure why the build is broken. Does AppVeyor not automatically restore NuGet packages?

Looks like that's actually the problem. The appveyor.xml file I have in this repo doesn't have a restore step. You'll need to add the following section:

``` yaml
before_build: 
  # Updates the version number in the .vsixmanifest and updates the AppVeyor build number to match
  - nuget restore VsixUtil.sln
```

**Comment by jcansdale** [comment]
Still trying to work out why the build is failing. 😕 

I notice there is a branch called `pr/6-split-into-separate-files-and-add-unit-tests`. Was that you or is this a GitHub thing?

**Comment by jcansdale** [comment]
Okay, so I need to fix the tests now. Yay for unit tests. 😄

**Pull request merged**